File: apps/api/app/schemas/execution.py
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class CreateExecutionRequest(BaseModel):
    mode: str = "IMMEDIATE"
    target_devices: List[str]
    ccm_payload: Dict[str, Any]


class ExecutionStepResponse(BaseModel):
    id: str
    step_name: str
    status: str
    command_text: Optional[str] = None
    output_log: Optional[str] = None


class ExecutionPlanResponse(BaseModel):
    id: str
    organization_id: str
    mode: str
    status: str
    risk_level: str = "SAFE"
    target_devices: List[str]
    driver_used: str = "Scrapli"
    steps: List[ExecutionStepResponse] = []
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    failure_reason: Optional[str] = None


class VerificationReportResponse(BaseModel):
    is_verified: bool
    target_host: str
    running_config_hash: str
    startup_config_hash: str
    diff_matched: bool
    verification_summary: str

File: packages/types/src/execution.ts
export enum ExecutionMode {
  DRY_RUN = 'DRY_RUN',
  PREVIEW_ONLY = 'PREVIEW_ONLY',
  IMMEDIATE = 'IMMEDIATE',
  SCHEDULED = 'SCHEDULED',
  MAINTENANCE_WINDOW = 'MAINTENANCE_WINDOW',
  BULK_EXECUTION = 'BULK_EXECUTION',
  CANARY_DEPLOYMENT = 'CANARY_DEPLOYMENT',
  PHASED_ROLLOUT = 'PHASED_ROLLOUT',
  EMERGENCY = 'EMERGENCY',
  ROLLBACK = 'ROLLBACK',
}

export enum ExecutionStatus {
  QUEUED = 'QUEUED',
  PREPARING = 'PREPARING',
  WAITING_APPROVAL = 'WAITING_APPROVAL',
  CONNECTING = 'CONNECTING',
  AUTHENTICATING = 'AUTHENTICATING',
  EXECUTING = 'EXECUTING',
  VERIFYING = 'VERIFYING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
  CANCELLED = 'CANCELLED',
  ROLLING_BACK = 'ROLLING_BACK',
  ROLLBACK_COMPLETED = 'ROLLBACK_COMPLETED',
}

export interface ExecutionStep {
  id: string;
  stepName: string;
  status: 'PENDING' | 'RUNNING' | 'SUCCESS' | 'FAILED';
  commandText?: string;
  outputLog?: string;
  executedAt?: string;
}

export interface NapalmVerificationReport {
  isVerified: boolean;
  runningConfigHash: string;
  startupConfigHash: string;
  diffMatched: boolean;
  interfaceStates: { name: string; operStatus: 'UP' | 'DOWN' }[];
}

export interface ExecutionPlanEntity {
  id: string;
  organizationId: string;
  initiatedBy: string;
  mode: ExecutionMode;
  status: ExecutionStatus;
  riskLevel: string;
  targetDevices: string[];
  driverUsed: 'Scrapli' | 'Netmiko (Fallback)';
  steps: ExecutionStep[];
  startedAt?: string;
  completedAt?: string;
  failureReason?: string;
  verificationReport?: NapalmVerificationReport;
}
